7_drivers_munge/src/GCM_driver_utils.R

```diff
@@ -401,3 +398,66 @@ validate_notaro_units_assumptions <- function(data_in) {
+#' `data_in` that is not `DateTime` or `time(day of year)`.
+convert_notaro_dates <- function(data_in) {
+  data_in %>%
+    # select(DateTime, `time(day of year)`, variable) %>%
+
+    ### Group data into the different big time chunks by finding and identifying ###
+
+    # This groups data at the points where 2000 is right before 2040, and 2040 is right before 2060
+    mutate(diff_days = abs(as.numeric(DateTime - lag(DateTime))/86400)) %>%
+    # Use a difference in days greater than one year to determine if a row is the
+    # start of a new time period
+    mutate(new_time_period = is.na(diff_days) | diff_days > 366) %>%
+    # Each new time period ticks the counter up by 1
+    mutate(new_time_period_i = cumsum(new_time_period)) %>%
+
+    ### For each time period + variable group, identify the earliest year in the period and use that to create a column of years ###
+
+    # Not able to just to `format(DateTime, "%Y")` to get the year because not all have a `DateTime` in the
+    # correct year, e.g. 1985 records have a DateTime of `1984-12-31 23:15:03`.
+
+    # First treat rows for the same variable and time period as a group
+    group_by(variable, new_time_period_i) %>%
+    # Determine the earliest year in the time period
+    mutate(earliest_year = as.numeric(format(min(DateTime), "%Y"))) %>% # mutate(year_1 = as.numeric(format(min(DateTime), "%Y"))) %>%
+    # Create groups of the years (`time(day of year)` will cycle through 0:364 for each
+    # year, so tick a counter up for every 0 in the column)
+    mutate(year_n = cumsum(`time(day of year)` == 0) - 1) %>%
+    mutate(year_i = earliest_year + year_n) %>%
+    ungroup() %>%
+
+    ### For each year + variable group, correctly convert to the DateTime field to a date ###
+
+    # The GCMs assume there are no leap years, but R automatically assumes leap years, so we need
+    # to implement a work around for that.
+
+    group_by(variable, new_time_period_i, year_i) %>%
+    # Start by creating a date and allow leap days to be added
+    mutate(date_initial = as.Date(sprintf("%s-01-01", year_i)) + `time(day of year)`) %>%
+    # If the date in the column is a leap day (Feb 29), the dates need to be adjusted by 1 day
+    # meaning the 59th day should be Mar 1 not Feb 29.
+    mutate(leap_adjustment = if_else(format(date_initial, "%m-%d") == "02-29", 1, 0)) %>%
+    # Use cumsum to make sure the leap_adjustment is used for every subsequent day of the year following a leap day
+    mutate(total_days_to_add = cumsum(leap_adjustment)) %>%
+    ungroup() %>% # Reset groups since they are no longer needed
+    # Using the initial date created, adjust the days to shift anything in a leap year that follows Feb 29
+    mutate(date_corrected = date_initial + total_days_to_add) %>%
+
+    # Keep only the variable and corrected date columns
+    select(date = date_corrected, matches("[0-9]+"), variable, statistic, units)
```

Comment:
Due to this PR requiring a lengthy download to run/verify, I'm not able to dig in and see what this `matches()` call is exposed to. But I think what this does is select date, variable, stat, units, and then any column that is _only_ numbers. Is that the case?

Reply:
Yes, correct! Any column that is _only_ numbers should be all the columns representing cells (since this is prior to pivoting). I suppose that comment is not comprehensive of what is kept by this `select()`